pkg/database/database.go

```diff
@@ -171,6 +176,11 @@ func (db *Database) OpenWithCacher(ctx context.Context, cacher cache.Cacher) err
 	// Enable auto-preloading.
 	rawDB = rawDB.Set("gorm:auto_preload", true)
 
+	// Prevent multiple simultaneous callback registrations due to a data race in
+	// gorm.
+	db.callbackLock.Lock()
```

Comment:
Are there multiple goroutines calling OpenWithCacher on the same DB handle?

If that's happening, that seems wrong. 

Instead of locking the callbacks, we should just do a quick lock/unlock see if the *Database has been opened or not.

Reply:
It's not the same db handle - see my comment in chat. It's the same callback handler. Gorm has a global `DefaultCallback` that is set on all `gorm.Open` calls. Thus the first call to `.Callback()` uses a shared global constant.

This only happens in tests because tests are the only thing that can call Open in parallel.

Reply:
saw that - thanks!

Reply:
so - should this be a global lock and not per-DB?

Reply:
Made it global, but I forget how Go works across packages with global vars.